source/features/useful-not-found-page.tsx

```diff
@@ -36,46 +38,156 @@ function parseCurrentURL(): string[] {
+		return;
+	}
+
+	// The object might have been deleted/moved
+	if (await addObjectStatus(bar)) {
+		return;
+	}
+
+	// Or 410 Gone
+	await addCommitHistoryLink(bar);
+}
+
+async function addObjectStatus(bar: Element): Promise<boolean> {
+	// Get the file path from the parts
+	const parts = parseCurrentURL();
+	const filePath = parts.slice(4).join('/');
+
+	// Get the last 2 commits that include the file
+	const previousCommitsIncludingFile = await api.v3(`commits?path=${filePath}&per_page=2`);
+	// The latest commit will be the first object in the array
+	if (previousCommitsIncludingFile[0]) {
+		// Get a list of changes that happened in the repo with this commit
+		const lastCommitInfo = await api.v3(`commits/${previousCommitsIncludingFile[0].sha as string}`);
+
+		// Check what happened to this file
+		const fileInfo = lastCommitInfo.files.find((file: Record<string, string>) => {
+			if (file.filename === filePath) {
+				return true;
+			}
+
+			if (file.status === 'renamed' && file.previous_filename === filePath) {
+				return true;
+			}
+
+			return false;
+		});
+
+		const commitAuthor = lastCommitInfo.author.login;
+		const commitTime = new Date(lastCommitInfo.commit.committer.date);
+		const commitSha = lastCommitInfo.sha;
+		const shortenedCommitSha = commitSha.slice(0, 8);
+		const commitUrl = lastCommitInfo.html_url;
+
+		const urlToCommitAuthorProfile = lastCommitInfo.author.html_url;
+		const urlToLastBlob = fileInfo.blob_url;
+
+		// If it was removed, tell the user
+		if (fileInfo.status === 'removed') {
+			bar.after(
+				<p className="container mt-4 text-center">
+					The file you are looking for was deleted/moved by <a href={urlToCommitAuthorProfile}>{commitAuthor}</a> {twas(commitTime.getTime())} with commit <a href={commitUrl}>{shortenedCommitSha}</a>.
+					<br/>
+					You can view the last version of the file <a href={urlToLastBlob}>here</a>.
+				</p>,
+			);
+
+			return true;
+		}
+
+		// If it was renamed, tell the user
+		if (fileInfo.status === 'renamed') {
+			bar.after(
+				<p className="container mt-4 text-center">
+					The file you are looking for was renamed by <a href={urlToCommitAuthorProfile}>{commitAuthor}</a> {twas(commitTime.getTime())} with commit <a href={commitUrl}>{shortenedCommitSha}</a>.
+					<br/>
+					You can find the renamed file <a href={urlToLastBlob}>here</a>.
+				</p>,
+			);
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+// If the object was deleted, link to the commit history
+async function addCommitHistoryLink(bar: Element): Promise<boolean> {
 	const parts = parseCurrentURL();
 	parts[2] = 'commits';
 	const url = '/' + parts.join('/');
 	if (await is404(location.origin + url)) {
-		return;
+		return false;
 	}
 
 	bar.after(
 		<p className="container mt-4 text-center">
 			See also the {getType()}’s <a href={url}>commit history</a>.
 		</p>,
 	);
+
+	return true;
+}
+
+async function addBranchStatus(bar: Element): Promise<boolean> {
```

Comment:
I just realized that this piece of information is already on screen and we don't need to fetch it again.

Reply:
I could change this to use the `is404` method instead of calling the API and comparing the current branch with existing branches. Does that sound ok?

Reply:
The info is already there, it just does not need to exist separately.

Reply:
I think it should piggyback on the existing calls to determine what happened to the current URL and display one of:

- branch was deleted
- file was deleted
- file was renamed/moved

Reply:
Refer to my other larger comment for a broader solution.